addOns/client/src/main/java/org/zaproxy/addon/client/ClientMap.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2022 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.client;
+
+import java.util.Comparator;
+import java.util.Locale;
+import javax.swing.tree.DefaultTreeModel;
+import javax.swing.tree.TreeNode;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@SuppressWarnings("serial")
+public class ClientMap extends SortedTreeModel {
+
+    private static final long serialVersionUID = 1L;
+    private static final Logger LOGGER = LogManager.getLogger(ClientMap.class);
+    private ClientNode root;
+
+    public ClientMap(ClientNode root) {
+        super(root);
+        this.root = root;
+    }
+
+    @Override
+    public ClientNode getRoot() {
+        return root;
+    }
+
+    public synchronized ClientNode getOrAddNode(String url, boolean storage) {
+        LOGGER.debug("getOrAddNode {}", url);
+        if (url == null) {
+            throw new IllegalArgumentException("The url parameter should not be null");
+        }
+        // Parse the URL into its component pieces
```

Comment:
I don't see anything wrong here but I feel like we must have this handling somewhere in core? 🤷‍♂️